### DISCUSSION TITLE
answermode early questions

### DISCUSSION THREAD

**Discussion by juha-h** [baresip/baresip#2829] [Q&A]
When `answermode=early-audio` is selected, baresip is replying with `180 Ringing` followed by `183 Session Progress`.   Why not just with `183 Session Progress`?

Is it possible to select which audio source baresip is using during `183 Session Progress` ?

Is this Wiki Accounts/answermode explanation correct
```
early-audio: The user has to answer the call. The device is not ringing, instead early audio will be requested from the peer.
```
since peer usually means the other party of the call?

**Comment by juha-h** [comment]
I made a test where config has
```
audio_source            gst,file:///home/jh/digits.wav 
```
and account has
```
;answermode=early-audio
```
When call comes in, caller still hears ringing sound, not `digits.wav`.  Why?

If account has 
```
;answermode=auto
```
call is answered and caller hears the digits.

**Reply by cspiel1** [reply]
This is correct.

If you expect to hear digits.wav, you have to configure this at the peer.
```
audio_source            gst,file:///home/jh/digits.wav
```

And this at the local baresip instance:
```
;answermode=early-audio
```

**Comment by cspiel1** [comment]
> When `answermode=early-audio` is selected, baresip is replying with `180 Ringing` followed by `183 Session Progress`. Why not just with `183 Session Progress`?

`call_accept()` sends the `180 Ringing` always. And the optional `call_progress()` sends the `183 Session Progress`. This is by design. Only the application (module, e.g. menu) decides if `call_progress()` should be called.

> 
> Is it possible to select which audio source baresip is using during `183 Session Progress` ?

No.

> 
> Is this Wiki Accounts/answermode explanation correct
> 
> ```
> early-audio: The user has to answer the call. The device is not ringing, instead early audio will be requested from the peer.
> ```
> 
> since peer usually means the other party of the call?

Yes, it is correct. And yes, the early-audio is requested from the other party of the call = peer.

**Comment by juha-h** [comment]
I have always thought that 183 response is used by the callee to play something to the caller, such as call recording or rate notice, that alllows the caller to terminate the call before it is answered.  See for example https://jazchye.com/index.php/2022/01/06/180-ringing-and-183-session-progress-in-sip/.

**Comment by juha-h** [comment]
Any more comments on this?  Looks like baresip early answermode behavior does not match with common practice when it does not play early audio from the callee. Can you point to some third party document where the current behavior is described?

**Comment by juha-h** [comment]
Sems, for example, supports two early audio (183) behaviors: (1) callee plays early audio after which it answers with 404 and (2) callee plays early audio after which the call starts to ring at the callee.  This makes lots of sense whereas I can't figure out when the current behavior would be useful.

**Comment by cspiel1** [comment]
Baresip does currently not allow to specify an audio file for early-media. There is only the `ausrc` that can be used. When Alice calls Bob and then Bob immediately streams the microphone to Alice without accepting the call, this would allow hidden surveillance.

We only use `early-video`. The doorstation Alice calls the operator Bob. Bob requests early-video (SIP 183 Session Progress) from Alice. The operator Bob sees who is standing at the door before accepting the call.

**Comment by juha-h** [comment]
Looks to me that you have invented your own 183 behavior for your own application, which contradicts how 183 normally works.
